### PR TITLE
ref(integrations): Refactor platform detection to composable framework definitions

### PR DESCRIPTION
**Pull request by jaydgoss**
## Summary

- Refactor flat framework detection into composable `FrameworkDef` / `DetectorRule` system
- Add three signal types: `path` (config file existence), `match_content` (regex on file content), `match_package` (dependency lookup in parsed manifests)
- Add `every` (AND) and `some` (OR) rule composition
- Add priority ranking (`sort` field) and supersession (e.g. Next.js supersedes React)
- Refactor file content fetching into a single batch pass to minimize API calls

No behavior change for existing detections, but the architecture now supports easy addition of new frameworks as data-only entries.

**Stack:**
- [PR 1](https://github.com/getsentry/sentry/pull/109699): Core detection + API endpoint
- **PR 2 (this):** Composable framework definitions refactor
- [PR 3](https://github.com/getsentry/sentry/pull/109701): Expanded coverage (98% of picker platforms)

### CODE REVIEW

File: src/sentry/integrations/github/platform_detection.py [hunk 1]
```diff
@@ -188,84 +354,118 @@ def _get_repo_file_content(
...
+            pkg = json.loads(content)
+            return _PackageManifest(
+                dependencies=set(pkg.get("dependencies", {}).keys()),
+                dev_dependencies=set(pkg.get("devDependencies", {}).keys()),
+            )
+        elif manifest_file == "composer.json":
+            composer = json.loads(content)
+            return _PackageManifest(
+                dependencies=set(composer.get("require", {}).keys()),
+                dev_dependencies=set(composer.get("require-dev", {}).keys()),
```
[the remark this answers is not in the document]
Comment (jaydgoss) [reply src/sentry/integrations/github/platform_detection.py]: Fixed in 4c7e1239ac2 — now uses `(pkg.get("dependencies") or {}).keys()` which handles both missing keys and explicit `null` values. Same fix applied to composer.json parsing. Tests added for both.

File: src/sentry/integrations/github/platform_detection.py [hunk 2]
```diff
@@ -69,108 +72,271 @@ class DetectedPlatform(TypedDict):
...
+    match_package: str  # Package name in package.json/composer.json deps
+
+
+class FrameworkDef(TypedDict, total=False):
+    platform: str  # Sentry platform ID, e.g. "javascript-nextjs"
+    sort: int  # Lower = higher priority (Vercel convention)
+    base_platform: str  # Language group, e.g. "javascript"
+    every: list[DetectorRule]  # ALL must match (AND)
+    some: list[DetectorRule]  # At least ONE must match (OR)
+    supersedes: list[str]  # Platform IDs this makes redundant
```
[the remark this answers is not in the document]
Comment (jaydgoss) [reply src/sentry/integrations/github/platform_detection.py]: Fixed in 4c7e1239ac2 — `FrameworkDef` now uses `total=True` with `NotRequired` on optional fields (`every`, `some`, `supersedes`). The required fields `platform`, `sort`, and `base_platform` are now enforced by the type checker.

File: src/sentry/integrations/github/platform_detection.py [hunk 3]
```diff
@@ -188,84 +354,118 @@ def _get_repo_file_content(
...
-    Refine a base platform (e.g. "python") into specific framework
-    platforms (e.g. "python-django") by reading manifest files.
+    try:
+        params: dict[str, str] = {}
+        if ref:
+            params["ref"] = ref
+        response = client.get(f"/repos/{repo}/contents", params=params)
+        return {item["name"] for item in response if item.get("type") == "file"}
+    except ApiError:
+        return set()
```
[the remark this answers is not in the document]
Comment (jaydgoss) [reply src/sentry/integrations/github/platform_detection.py]: Fixed in c6b632ec055 — now catches `(ApiError, AttributeError, KeyError, TypeError)`. Handles malformed items (missing `"name"` key) and non-list responses. Tests added for both edge cases.

File: src/sentry/integrations/github/platform_detection.py [hunk 4]
```diff
@@ -69,106 +72,277 @@ class DetectedPlatform(TypedDict):
...
+            {"path": "pom.xml", "match_content": r"spring-framework"},
+        ],
+    },
+    # --- Go ---
+    {
+        "platform": "go-echo",
+        "sort": 20,
+        "base_platform": "go",
+        "some": [
+            {"path": "go.mod", "match_content": r"(?i)\becho\b"},
```
Comment (jaydgoss) [inline src/sentry/integrations/github/platform_detection.py:306]: Note: These Go patterns (`\becho\b`, `\bgin\b`, `\bfiber\b`) are loose enough to match comments or unrelated text in go.mod.

PR 3 in this stack (#109701) replaces these with full module paths (`github.com/labstack/echo`, `github.com/gin-gonic/gin`, `github.com/gofiber/fiber`) which are precise and don't false-positive.

File: src/sentry/integrations/github/platform_detection.py [hunk 5]
```diff
@@ -69,106 +72,277 @@ class DetectedPlatform(TypedDict):
...
+    some: NotRequired[list[DetectorRule]]  # At least ONE must match (OR)
+    supersedes: NotRequired[list[str]]  # Platform IDs this makes redundant
+
+
+# Each framework is a self-contained definition with composable detector rules.
+#
+# The `sort` field controls priority for onboarding recommendations.
+# Lower sort = higher priority. Converted to `priority = 100 - sort` in output.
+#
+#   sort=1      Meta-frameworks         Next.js, Remix, Nuxt, SvelteKit
```
Comment (wedamija) [inline src/sentry/integrations/github/platform_detection.py:99]: Why do we prioritise js frameworks over everything else?

File: src/sentry/integrations/github/platform_detection.py [hunk 6]
```diff
@@ -69,106 +72,277 @@ class DetectedPlatform(TypedDict):
...
+    priority: int  # Higher = more relevant for onboarding
+
+
+class DetectorRule(TypedDict, total=False):
+    path: str  # File must exist in root directory
+    match_content: str  # Regex pattern to match in file content (requires path)
+    match_package: str  # Package name in package.json/composer.json deps
+
+
+class FrameworkDef(TypedDict):
```
Comment (wedamija) [inline src/sentry/integrations/github/platform_detection.py:84]: Would this be better off as a dataclass? It might make the definitions a little more concise, and we could set default for lists and so on so that we don't have to do `get("every", [])` in a bunch of places

File: src/sentry/integrations/github/platform_detection.py [hunk 7]
```diff
@@ -188,84 +362,118 @@ def _get_repo_file_content(
...
+def _rule_matches(
+    rule: DetectorRule,
+    root_files: set[str],
+    file_contents: dict[str, str],
+    package_manifest: _PackageManifest | None,
+) -> bool:
+    """Evaluate a single detector rule against repository state."""
+    if "match_package" in rule:
+        if package_manifest is None:
+            return False
```
Comment (wedamija) [inline src/sentry/integrations/github/platform_detection.py]: Feels like maybe this could be part of _package_in_manifest?

File: src/sentry/integrations/github/platform_detection.py [hunk 8]
```diff
@@ -188,84 +362,118 @@ def _get_repo_file_content(
...
+    some: Sequence[DetectorRule] = fw.get("some", [])
 
-    return unique
+    if not every and not some:
+        return False
+
+    every_pass = (
+        all(_rule_matches(r, root_files, file_contents, package_manifest) for r in every)
+        if every
+        else True
```
Comment (wedamija) [inline src/sentry/integrations/github/platform_detection.py]: The if/else isn't necessary here. `all((a for a in []))` returns True

File: src/sentry/integrations/github/platform_detection.py [hunk 9]
```diff
@@ -188,84 +362,118 @@ def _get_repo_file_content(
...
+
+    return every_pass and some_pass
+
+
+def _apply_supersession(results: list[DetectedPlatform]) -> list[DetectedPlatform]:
+    """Remove platforms that are superseded by more specific ones.
+
+    e.g. if Next.js is detected, React is redundant since Next.js includes it.
+    """
+    detected_ids = {r["platform"] for r in results}
```
Comment (wedamija) [inline src/sentry/integrations/github/platform_detection.py]: Nit: `platforms/platform_ids`

File: src/sentry/integrations/github/platform_detection.py (hunk 5 above)
Comment (jaydgoss) [reply src/sentry/integrations/github/platform_detection.py:99]: ~~Sort controls ordering across all detected frameworks in the result set, not just within a language.~~ Update: this was a valid concern. Previously sort was the primary ranking factor across all languages, so Next.js (sort=1) would outrank Django (sort=10) even in a Python-majority repo.

Fixed in 004dc54 -- results now sort by `(bytes, priority)` so language majority is the primary factor. Sort only controls ordering within a language group.

Example with Python=80k, JS=30k:
- Before: Next.js (priority=99, 30k) > Django (priority=90, 80k)
- After: Django (priority=90, 80k) > Next.js (priority=99, 30k)

The tiers (which now only matter within a language):

- sort=1: Meta-frameworks (Next.js, Remix, Nuxt, SvelteKit)
- sort=10: Primary frameworks (Django, Rails, Laravel, Spring Boot)
- sort=20: Secondary frameworks (Flask, Go frameworks, Symfony)
- sort=30: UI / general (React, Vue, Angular, Svelte, Starlette)
- sort=40: Server frameworks (Express, Koa)
- sort=60: Utilities / background (Celery)

File: src/sentry/integrations/github/platform_detection.py (hunk 6 above)
Comment (jaydgoss) [reply src/sentry/integrations/github/platform_detection.py:84]: Good call. Filed as a follow-up since it touches every framework definition and the evaluation logic. The main wins would be default values for optional lists (no more .get("every", [])) and better IDE support.

Fixed the other nits from this review in 21af685:
- Folded None-manifest check into _package_in_manifest
- Simplified _framework_matches (all([]) returns True)
- Renamed detected_ids to platform_ids

File: src/sentry/integrations/github/platform_detection.py (hunk 7 above)
Comment (jaydgoss) [reply src/sentry/integrations/github/platform_detection.py]: Done in 21af685. _package_in_manifest now accepts None and returns False, so _rule_matches just calls through directly.

File: src/sentry/integrations/github/platform_detection.py (hunk 8 above)
Comment (jaydgoss) [reply src/sentry/integrations/github/platform_detection.py]: Fixed in 21af685. Simplified to just all(...) since all([]) returns True. Kept the some guard since any([]) returns False.

File: src/sentry/integrations/github/platform_detection.py (hunk 9 above)
Comment (jaydgoss) [reply src/sentry/integrations/github/platform_detection.py]: Fixed in 21af685.

**Review by wedamija — APPROVED** [review]

File: src/sentry/integrations/github/platform_detection.py [hunk 10]
```diff
@@ -276,37 +481,84 @@ def detect_platforms(
...
+    for base_platform in active_platforms:
+        for fw in _FRAMEWORKS_BY_PLATFORM.get(base_platform, []):
+            for rule in [*fw.get("every", []), *fw.get("some", [])]:
+                path = rule.get("path")
+                if path and "match_content" in rule and path in root_files:
+                    needed_paths.add(path)
+
+    # Fetch file contents in one pass
+    file_contents: dict[str, str] = {}
+    for path in needed_paths:
```
[the remark this answers is not in the document]
Comment (jaydgoss) [reply src/sentry/integrations/github/platform_detection.py:525]: Good catch. Fixed in ce98f6bc674.

`_get_root_file_names` (now `_get_root_entries` in the next PR) returns `None` on API failure instead of empty set. `detect_platforms` then skips the root-listing gates and tries fetching manifest/content files individually. Path-only rules return `False` when the listing is unavailable since we can't confirm file existence, but content-based and package-based detection still works.

Added a test (`test_root_listing_failure_still_detects_frameworks_via_manifest`) that verifies Django is detected from requirements.txt even when the root contents API returns 500.
